{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f1c45261-45d0-4319-86f3-9dac23b9b61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from pyfaidx import Fasta\n",
    "\n",
    "from tqdm.notebook import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef9a0618-fca1-4c67-a6de-e2a29bf0aa30",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_226562/1922461105.py:1: DtypeWarning: Columns (14) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(\"data/ClinVar/SNV_variants.csv\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>150829393</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>4</td>\n",
       "      <td>NC_000015.9</td>\n",
       "      <td>15</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>150829393</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>4</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>84799209</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>267606829</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>NC_000011.9</td>\n",
       "      <td>11</td>\n",
       "      <td>126145284</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>267606829</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>5</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>126275389</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15045</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.1289A&gt;G (p.Asn430Ser)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>267606830</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>6</td>\n",
       "      <td>NC_000011.9</td>\n",
       "      <td>11</td>\n",
       "      <td>126147412</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835686</th>\n",
       "      <td>1952238</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001844.5(COL2A1):c.1502G&gt;A (p.Gly501Glu)</td>\n",
       "      <td>1280</td>\n",
       "      <td>COL2A1</td>\n",
       "      <td>HGNC:2200</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Spondyloepiphyseal dysplasia congenita</td>\n",
       "      <td>de novo</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895438</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>47986361</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835687</th>\n",
       "      <td>1952239</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000263.4(NAGLU):c.1978A&gt;G (p.Asn660Asp)</td>\n",
       "      <td>4669</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>HGNC:7632</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Mucopolysaccharidosis, MPS-III-B</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1895439</td>\n",
       "      <td>NC_000017.10</td>\n",
       "      <td>17</td>\n",
       "      <td>40696002</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835688</th>\n",
       "      <td>1952239</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000263.4(NAGLU):c.1978A&gt;G (p.Asn660Asp)</td>\n",
       "      <td>4669</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>HGNC:7632</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Mucopolysaccharidosis, MPS-III-B</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895439</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>42543984</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835689</th>\n",
       "      <td>1952240</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000435.3(NOTCH3):c.2903C&gt;A (p.Ser968Ter)</td>\n",
       "      <td>4854</td>\n",
       "      <td>NOTCH3</td>\n",
       "      <td>HGNC:7883</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Cerebral arteriopathy, autosomal dominant, wit...</td>\n",
       "      <td>paternal</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1895440</td>\n",
       "      <td>NC_000019.9</td>\n",
       "      <td>19</td>\n",
       "      <td>15291863</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835690</th>\n",
       "      <td>1952240</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000435.3(NOTCH3):c.2903C&gt;A (p.Ser968Ter)</td>\n",
       "      <td>4854</td>\n",
       "      <td>NOTCH3</td>\n",
       "      <td>HGNC:7883</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Cerebral arteriopathy, autosomal dominant, wit...</td>\n",
       "      <td>paternal</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895440</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>15181052</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2835691 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #AlleleID                       Type  \\\n",
       "0            15043  single nucleotide variant   \n",
       "1            15043  single nucleotide variant   \n",
       "2            15044  single nucleotide variant   \n",
       "3            15044  single nucleotide variant   \n",
       "4            15045  single nucleotide variant   \n",
       "...            ...                        ...   \n",
       "2835686    1952238  single nucleotide variant   \n",
       "2835687    1952239  single nucleotide variant   \n",
       "2835688    1952239  single nucleotide variant   \n",
       "2835689    1952240  single nucleotide variant   \n",
       "2835690    1952240  single nucleotide variant   \n",
       "\n",
       "                                                 Name  GeneID GeneSymbol  \\\n",
       "0        NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "1        NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "2         NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "3         NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "4        NM_017547.4(FOXRED1):c.1289A>G (p.Asn430Ser)   55572    FOXRED1   \n",
       "...                                               ...     ...        ...   \n",
       "2835686   NM_001844.5(COL2A1):c.1502G>A (p.Gly501Glu)    1280     COL2A1   \n",
       "2835687    NM_000263.4(NAGLU):c.1978A>G (p.Asn660Asp)    4669      NAGLU   \n",
       "2835688    NM_000263.4(NAGLU):c.1978A>G (p.Asn660Asp)    4669      NAGLU   \n",
       "2835689   NM_000435.3(NOTCH3):c.2903C>A (p.Ser968Ter)    4854     NOTCH3   \n",
       "2835690   NM_000435.3(NOTCH3):c.2903C>A (p.Ser968Ter)    4854     NOTCH3   \n",
       "\n",
       "            HGNC_ID    ClinicalSignificance  RS# (dbSNP)  \\\n",
       "0        HGNC:28986  Uncertain significance    150829393   \n",
       "1        HGNC:28986  Uncertain significance    150829393   \n",
       "2        HGNC:26927              Pathogenic    267606829   \n",
       "3        HGNC:26927              Pathogenic    267606829   \n",
       "4        HGNC:26927              Pathogenic    267606830   \n",
       "...             ...                     ...          ...   \n",
       "2835686   HGNC:2200       Likely pathogenic           -1   \n",
       "2835687   HGNC:7632       Likely pathogenic           -1   \n",
       "2835688   HGNC:7632       Likely pathogenic           -1   \n",
       "2835689   HGNC:7883       Likely pathogenic           -1   \n",
       "2835690   HGNC:7883       Likely pathogenic           -1   \n",
       "\n",
       "                                             PhenotypeList            Origin  \\\n",
       "0                                Galloway-Mowat syndrome 1          germline   \n",
       "1                                Galloway-Mowat syndrome 1          germline   \n",
       "2        Mitochondrial complex 1 deficiency, nuclear ty...  germline;unknown   \n",
       "3        Mitochondrial complex 1 deficiency, nuclear ty...  germline;unknown   \n",
       "4        Mitochondrial complex 1 deficiency, nuclear ty...          germline   \n",
       "...                                                    ...               ...   \n",
       "2835686             Spondyloepiphyseal dysplasia congenita           de novo   \n",
       "2835687                   Mucopolysaccharidosis, MPS-III-B          germline   \n",
       "2835688                   Mucopolysaccharidosis, MPS-III-B          germline   \n",
       "2835689  Cerebral arteriopathy, autosomal dominant, wit...          paternal   \n",
       "2835690  Cerebral arteriopathy, autosomal dominant, wit...          paternal   \n",
       "\n",
       "        OriginSimple Assembly  VariationID ChromosomeAccession Chromosome  \\\n",
       "0           germline   GRCh37            4         NC_000015.9         15   \n",
       "1           germline   GRCh38            4        NC_000015.10         15   \n",
       "2           germline   GRCh37            5         NC_000011.9         11   \n",
       "3           germline   GRCh38            5        NC_000011.10         11   \n",
       "4           germline   GRCh37            6         NC_000011.9         11   \n",
       "...              ...      ...          ...                 ...        ...   \n",
       "2835686     germline   GRCh38      1895438        NC_000012.12         12   \n",
       "2835687     germline   GRCh37      1895439        NC_000017.10         17   \n",
       "2835688     germline   GRCh38      1895439        NC_000017.11         17   \n",
       "2835689     germline   GRCh37      1895440         NC_000019.9         19   \n",
       "2835690     germline   GRCh38      1895440        NC_000019.10         19   \n",
       "\n",
       "         PositionVCF ReferenceAlleleVCF AlternateAlleleVCF  \n",
       "0           85342440                  G                  A  \n",
       "1           84799209                  G                  A  \n",
       "2          126145284                  C                  T  \n",
       "3          126275389                  C                  T  \n",
       "4          126147412                  A                  G  \n",
       "...              ...                ...                ...  \n",
       "2835686     47986361                  C                  T  \n",
       "2835687     40696002                  A                  G  \n",
       "2835688     42543984                  A                  G  \n",
       "2835689     15291863                  G                  T  \n",
       "2835690     15181052                  G                  T  \n",
       "\n",
       "[2835691 rows x 18 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"data/ClinVar/SNV_variants.csv\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5e6b09e8-22d0-4447-946a-8d0880fe509c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['15', '11', '14', '6', '2', '20', '10', '19', '16', '22', '1', '7',\n",
       "       '8', '9', '21', '5', '4', '17', '18', '3', '12', 'na', '13', 'MT',\n",
       "       'Y', 'X', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,\n",
       "       17, 18, 19, 20, 21, 22, 'Un'], dtype=object)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\"Chromosome\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "533200d1-c50e-4442-ab76-a3698dad9f31",
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_chromosomes = [str(i) for i in range(1, 23)] +[\"X\", \"Y\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fb1cce05-7fc0-4864-bbef-17346a4c19fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "buffer_size = 7\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4359956-5bda-40b7-8d7f-62a1c2fc3809",
   "metadata": {},
   "source": [
    "# Process hg19 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c5af4ebb-4878-4f4c-9aac-ef3e6465e327",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "odict_keys(['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX', 'chrY', 'chrM', 'GL877870.2', 'GL877872.1', 'GL383535.1', 'JH159133.1', 'KB663606.1', 'KB021647.1', 'KE332497.1', 'JH159131.1', 'GL949745.1', 'JH720447.1', 'GL582968.1', 'JH720446.1', 'JH591181.2', 'JH720443.2', 'JH159134.2', 'JH636052.4', 'JH636053.3', 'JH636054.1', 'JH636056.1', 'JH636058.1', 'JH636057.1', 'KE332505.1', 'JH720451.1', 'JH720452.1', 'JH720453.1', 'JH720454.3', 'JH159136.1', 'JH806587.1', 'JH806588.1', 'JH806589.1', 'JH806590.2', 'JH806591.1', 'JH806592.1', 'JH806593.1', 'JH806594.1', 'JH806595.1', 'JH806596.1', 'JH806597.1', 'JH720448.1', 'JH806598.1', 'JH806599.1', 'JH806600.2', 'JH806601.1', 'JH806602.1', 'JH806573.1', 'JH806574.2', 'JH806575.1', 'JH806580.1', 'JH806583.1', 'JH806584.1', 'JH806585.1', 'JH806603.1', 'JH159150.3', 'GL582969.1', 'JH806577.1', 'JH806578.1', 'JH806579.1', 'JH159137.1', 'KE332502.1', 'KB021645.1', 'KB663607.2', 'KE332500.1', 'KE332496.1', 'GL383558.1', 'GL582976.1', 'GL383523.1', 'KE332498.1', 'GL949743.1', 'KE332506.1', 'GL383536.1', 'JH591184.1', 'JH636061.1', 'JH806576.1', 'GL383524.1', 'GL582973.1', 'JH159138.1', 'KB021648.1', 'JH159139.1', 'JH159140.1', 'JH591182.1', 'JH159132.1', 'JH720449.1', 'JH591183.1', 'JH720444.2', 'JH159141.2', 'KB663604.1', 'JH720455.1', 'KB021646.2', 'JH159142.2', 'JH159143.1', 'JH806582.2', 'JH159135.2', 'KE332499.1', 'GL877877.2', 'JH806586.1', 'GL582979.2', 'KB663605.1', 'GL582975.1', 'GL949744.1', 'GL383543.1', 'GL877871.1', 'GL582967.1', 'JH159149.1', 'GL582977.2', 'GL582970.1', 'GL383559.2', 'JH159144.1', 'JH591186.1', 'GL383560.1', 'GL339450.1', 'GL582971.1', 'GL582974.1', 'JH806581.1', 'JH636060.1', 'JH591185.1', 'JH159145.1', 'GL877873.1', 'KB663608.1', 'GL582972.1', 'KB663603.1', 'KE332495.1', 'JH636059.1', 'JH720445.1', 'KE332501.1', 'GL383561.2', 'GL949741.1', 'GL383562.1', 'GL383525.1', 'GL383544.1', 'GL383548.1', 'GL383537.1', 'GL383538.1', 'GL383516.1', 'GL383517.1', 'GL383545.1', 'GL383546.1', 'GL383547.1', 'GL877875.1', 'GL383549.1', 'GL383550.1', 'GL383551.1', 'GL877876.1', 'GL383552.1', 'GL383553.2', 'GL383554.1', 'GL383555.1', 'GL383556.1', 'GL383557.1', 'GL000258.1', 'GL383563.2', 'GL383564.1', 'GL383565.1', 'GL383566.1', 'JH159146.1', 'JH159147.1', 'JH159148.1', 'GL383567.1', 'GL383568.1', 'GL383569.1', 'GL383570.1', 'GL383571.1', 'GL383572.1', 'GL949746.1', 'GL949747.1', 'GL949748.1', 'GL949749.1', 'GL949750.1', 'GL949751.1', 'GL949752.1', 'GL949753.1', 'GL383573.1', 'GL383574.1', 'GL383575.2', 'GL383576.1', 'GL383518.1', 'GL383519.1', 'GL383520.1', 'GL383577.1', 'GL383578.1', 'GL383579.1', 'GL383580.1', 'GL383581.1', 'GL383582.2', 'GL383583.1', 'KB663609.1', 'GL383521.1', 'GL383522.1', 'GL582966.2', 'JH636055.1', 'GL383526.1', 'GL000257.1', 'GL383527.1', 'GL383528.1', 'GL383529.1', 'GL339449.2', 'GL383530.1', 'GL383531.1', 'GL383532.1', 'GL949742.1', 'GL383533.1', 'KB021644.1', 'GL000250.1', 'GL000251.1', 'GL000252.1', 'GL000253.1', 'GL000254.1', 'GL000255.1', 'GL000256.1', 'GL383534.2', 'GL383539.1', 'GL383540.1', 'GL383541.1', 'GL383542.1', 'GL000191.1', 'GL000192.1', 'GL000193.1', 'GL000194.1', 'GL000195.1', 'GL000196.1', 'GL000197.1', 'GL000198.1', 'GL000199.1', 'GL000200.1', 'GL000201.1', 'GL000202.1', 'GL000203.1', 'GL000204.1', 'GL000205.1', 'GL000206.1', 'GL000207.1', 'GL000208.1', 'GL000209.1', 'GL000210.1', 'GL000211.1', 'GL000212.1', 'GL000213.1', 'GL000214.1', 'GL000215.1', 'GL000216.1', 'GL000217.1', 'GL000218.1', 'GL000219.1', 'GL000220.1', 'GL000221.1', 'GL000222.1', 'GL000223.1', 'GL000224.1', 'GL000225.1', 'GL000226.1', 'GL000227.1', 'GL000228.1', 'GL000229.1', 'GL000230.1', 'GL000231.1', 'GL000232.1', 'GL000233.1', 'GL000234.1', 'GL000235.1', 'GL000236.1', 'GL000237.1', 'GL000238.1', 'GL000239.1', 'GL000240.1', 'GL000241.1', 'GL000242.1', 'GL000243.1', 'GL000244.1', 'GL000245.1', 'GL000246.1', 'GL000247.1', 'GL000248.1', 'GL000249.1'])"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg_19_genome = Fasta('data/reference_genomes/GRCh37.p13/GRCh37.p13.genome.fa')\n",
    "hg_19_genome.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "44ca93f1-0c89-429d-abab-5cbd5ac9a3d5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>150829393</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>4</td>\n",
       "      <td>NC_000015.9</td>\n",
       "      <td>15</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>267606829</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>5</td>\n",
       "      <td>NC_000011.9</td>\n",
       "      <td>11</td>\n",
       "      <td>126145284</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15045</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.1289A&gt;G (p.Asn430Ser)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>267606830</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>6</td>\n",
       "      <td>NC_000011.9</td>\n",
       "      <td>11</td>\n",
       "      <td>126147412</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>15046</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_025152.3(NUBPL):c.166G&gt;A (p.Gly56Arg)</td>\n",
       "      <td>80224</td>\n",
       "      <td>NUBPL</td>\n",
       "      <td>HGNC:20278</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>200401432</td>\n",
       "      <td>not provided|Inborn genetic diseases|Mitochond...</td>\n",
       "      <td>germline;paternal</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>214885</td>\n",
       "      <td>NC_000014.8</td>\n",
       "      <td>14</td>\n",
       "      <td>32031331</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>15048</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000410.4(HFE):c.845G&gt;A (p.Cys282Tyr)</td>\n",
       "      <td>3077</td>\n",
       "      <td>HFE</td>\n",
       "      <td>HGNC:4886</td>\n",
       "      <td>Conflicting interpretations of pathogenicity; ...</td>\n",
       "      <td>1800562</td>\n",
       "      <td>Hemochromatosis type 1|Porphyria cutanea tarda...</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>9</td>\n",
       "      <td>NC_000006.11</td>\n",
       "      <td>6</td>\n",
       "      <td>26093141</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835681</th>\n",
       "      <td>1952235</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000152.5(GAA):c.844G&gt;C (p.Asp282His)</td>\n",
       "      <td>2548</td>\n",
       "      <td>GAA</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Glycogen storage disease, type II</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1895435</td>\n",
       "      <td>NC_000017.10</td>\n",
       "      <td>17</td>\n",
       "      <td>78081507</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835683</th>\n",
       "      <td>1952237</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_018706.7(DHTKD1):c.2546A&gt;T (p.Glu849Val)</td>\n",
       "      <td>55526</td>\n",
       "      <td>DHTKD1</td>\n",
       "      <td>HGNC:23537</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>778723208</td>\n",
       "      <td>Charcot-Marie-Tooth disease axonal type 2Q</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1895437</td>\n",
       "      <td>NC_000010.10</td>\n",
       "      <td>10</td>\n",
       "      <td>12160891</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835685</th>\n",
       "      <td>1952238</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001844.5(COL2A1):c.1502G&gt;A (p.Gly501Glu)</td>\n",
       "      <td>1280</td>\n",
       "      <td>COL2A1</td>\n",
       "      <td>HGNC:2200</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Spondyloepiphyseal dysplasia congenita</td>\n",
       "      <td>de novo</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1895438</td>\n",
       "      <td>NC_000012.11</td>\n",
       "      <td>12</td>\n",
       "      <td>48380144</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835687</th>\n",
       "      <td>1952239</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000263.4(NAGLU):c.1978A&gt;G (p.Asn660Asp)</td>\n",
       "      <td>4669</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>HGNC:7632</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Mucopolysaccharidosis, MPS-III-B</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1895439</td>\n",
       "      <td>NC_000017.10</td>\n",
       "      <td>17</td>\n",
       "      <td>40696002</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835689</th>\n",
       "      <td>1952240</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000435.3(NOTCH3):c.2903C&gt;A (p.Ser968Ter)</td>\n",
       "      <td>4854</td>\n",
       "      <td>NOTCH3</td>\n",
       "      <td>HGNC:7883</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Cerebral arteriopathy, autosomal dominant, wit...</td>\n",
       "      <td>paternal</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>1895440</td>\n",
       "      <td>NC_000019.9</td>\n",
       "      <td>19</td>\n",
       "      <td>15291863</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1417607 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #AlleleID                       Type  \\\n",
       "0            15043  single nucleotide variant   \n",
       "2            15044  single nucleotide variant   \n",
       "4            15045  single nucleotide variant   \n",
       "6            15046  single nucleotide variant   \n",
       "8            15048  single nucleotide variant   \n",
       "...            ...                        ...   \n",
       "2835681    1952235  single nucleotide variant   \n",
       "2835683    1952237  single nucleotide variant   \n",
       "2835685    1952238  single nucleotide variant   \n",
       "2835687    1952239  single nucleotide variant   \n",
       "2835689    1952240  single nucleotide variant   \n",
       "\n",
       "                                                 Name  GeneID GeneSymbol  \\\n",
       "0        NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "2         NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "4        NM_017547.4(FOXRED1):c.1289A>G (p.Asn430Ser)   55572    FOXRED1   \n",
       "6            NM_025152.3(NUBPL):c.166G>A (p.Gly56Arg)   80224      NUBPL   \n",
       "8             NM_000410.4(HFE):c.845G>A (p.Cys282Tyr)    3077        HFE   \n",
       "...                                               ...     ...        ...   \n",
       "2835681       NM_000152.5(GAA):c.844G>C (p.Asp282His)    2548        GAA   \n",
       "2835683   NM_018706.7(DHTKD1):c.2546A>T (p.Glu849Val)   55526     DHTKD1   \n",
       "2835685   NM_001844.5(COL2A1):c.1502G>A (p.Gly501Glu)    1280     COL2A1   \n",
       "2835687    NM_000263.4(NAGLU):c.1978A>G (p.Asn660Asp)    4669      NAGLU   \n",
       "2835689   NM_000435.3(NOTCH3):c.2903C>A (p.Ser968Ter)    4854     NOTCH3   \n",
       "\n",
       "            HGNC_ID                               ClinicalSignificance  \\\n",
       "0        HGNC:28986                             Uncertain significance   \n",
       "2        HGNC:26927                                         Pathogenic   \n",
       "4        HGNC:26927                                         Pathogenic   \n",
       "6        HGNC:20278                             Uncertain significance   \n",
       "8         HGNC:4886  Conflicting interpretations of pathogenicity; ...   \n",
       "...             ...                                                ...   \n",
       "2835681   HGNC:4065                                  Likely pathogenic   \n",
       "2835683  HGNC:23537                             Uncertain significance   \n",
       "2835685   HGNC:2200                                  Likely pathogenic   \n",
       "2835687   HGNC:7632                                  Likely pathogenic   \n",
       "2835689   HGNC:7883                                  Likely pathogenic   \n",
       "\n",
       "         RS# (dbSNP)                                      PhenotypeList  \\\n",
       "0          150829393                          Galloway-Mowat syndrome 1   \n",
       "2          267606829  Mitochondrial complex 1 deficiency, nuclear ty...   \n",
       "4          267606830  Mitochondrial complex 1 deficiency, nuclear ty...   \n",
       "6          200401432  not provided|Inborn genetic diseases|Mitochond...   \n",
       "8            1800562  Hemochromatosis type 1|Porphyria cutanea tarda...   \n",
       "...              ...                                                ...   \n",
       "2835681           -1                  Glycogen storage disease, type II   \n",
       "2835683    778723208         Charcot-Marie-Tooth disease axonal type 2Q   \n",
       "2835685           -1             Spondyloepiphyseal dysplasia congenita   \n",
       "2835687           -1                   Mucopolysaccharidosis, MPS-III-B   \n",
       "2835689           -1  Cerebral arteriopathy, autosomal dominant, wit...   \n",
       "\n",
       "                    Origin OriginSimple Assembly  VariationID  \\\n",
       "0                 germline     germline   GRCh37            4   \n",
       "2         germline;unknown     germline   GRCh37            5   \n",
       "4                 germline     germline   GRCh37            6   \n",
       "6        germline;paternal     germline   GRCh37       214885   \n",
       "8         germline;unknown     germline   GRCh37            9   \n",
       "...                    ...          ...      ...          ...   \n",
       "2835681           germline     germline   GRCh37      1895435   \n",
       "2835683           germline     germline   GRCh37      1895437   \n",
       "2835685            de novo     germline   GRCh37      1895438   \n",
       "2835687           germline     germline   GRCh37      1895439   \n",
       "2835689           paternal     germline   GRCh37      1895440   \n",
       "\n",
       "        ChromosomeAccession Chromosome  PositionVCF ReferenceAlleleVCF  \\\n",
       "0               NC_000015.9         15     85342440                  G   \n",
       "2               NC_000011.9         11    126145284                  C   \n",
       "4               NC_000011.9         11    126147412                  A   \n",
       "6               NC_000014.8         14     32031331                  G   \n",
       "8              NC_000006.11          6     26093141                  G   \n",
       "...                     ...        ...          ...                ...   \n",
       "2835681        NC_000017.10         17     78081507                  G   \n",
       "2835683        NC_000010.10         10     12160891                  A   \n",
       "2835685        NC_000012.11         12     48380144                  C   \n",
       "2835687        NC_000017.10         17     40696002                  A   \n",
       "2835689         NC_000019.9         19     15291863                  G   \n",
       "\n",
       "        AlternateAlleleVCF  \n",
       "0                        A  \n",
       "2                        T  \n",
       "4                        G  \n",
       "6                        A  \n",
       "8                        A  \n",
       "...                    ...  \n",
       "2835681                  C  \n",
       "2835683                  T  \n",
       "2835685                  T  \n",
       "2835687                  G  \n",
       "2835689                  T  \n",
       "\n",
       "[1417607 rows x 18 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg19_data = data[data[\"Assembly\"]==\"GRCh37\"]\n",
    "hg19_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ff159d8e-9477-4ac3-a1a4-b7f38cd6f454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['15', '11', '14', '6', '2', '20', '10', '19', '16', '22', '1', '7',\n",
       "       '8', '9', '21', '5', '4', '17', '18', '3', '12', '13', 'MT', 'Y',\n",
       "       'X', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,\n",
       "       19, 20, 21, 22, 'Un'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg19_data[\"Chromosome\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4225afd6-2aaa-419f-8e82-82f7b9094467",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e28e6b36c7d649a2a577f0ae84314d5a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hg19_processed_data = []\n",
    "for i, row in tqdm(hg19_data.iterrows()):\n",
    "    if row[\"Chromosome\"] not in valid_chromosomes:\n",
    "        continue\n",
    "        \n",
    "    ix = row[\"PositionVCF\"]-1\n",
    "    \n",
    "    if not hg_19_genome[\"chr\"+row[\"Chromosome\"]][ix]==row.ReferenceAlleleVCF:\n",
    "        continue\n",
    "    # assert hg_19_genome[\"chr\"+row[\"Chromosome\"]][ix]==row.ReferenceAlleleVCF or hg_19_genome[\"chr\"+row[\"Chromosome\"]][ix]==\"N\"\n",
    "    \n",
    "    preceding_sequence = hg_19_genome[\"chr\"+row[\"Chromosome\"]][ix-buffer_size:ix]\n",
    "    following_sequence = hg_19_genome[\"chr\"+row[\"Chromosome\"]][ix+1:ix+buffer_size+1]\n",
    "    hg19_processed_data.append({\"ReferenceSequence\": str(preceding_sequence).upper()+row.ReferenceAlleleVCF+str(following_sequence).upper(), \n",
    "                           \"MutatedSequence\": str(preceding_sequence).upper()+row.AlternateAlleleVCF+str(following_sequence).upper(),\n",
    "                          \"ClinicalSignificance\": row.ClinicalSignificance, \"OriginSimple\": row.OriginSimple, \"GeneSymbol\": row.GeneSymbol, \"HGNC_ID\": row.HGNC_ID, \"PhenotypeList\": row.PhenotypeList,\n",
    "                           \"RS# (dbSNP)\": row[\"RS# (dbSNP)\"]\n",
    "                          })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3c4fb23d-81ea-4bf8-848f-17d5d194d845",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ReferenceSequence</th>\n",
       "      <th>MutatedSequence</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CACCTGCGGGTGAGT</td>\n",
       "      <td>CACCTGCAGGTGAGT</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>150829393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGTCCAGTCCTT</td>\n",
       "      <td>AAAGGTCTAGTCCTT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>267606829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTTGTCAACATGTAC</td>\n",
       "      <td>GTTGTCAGCATGTAC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>267606830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTCCCGAGGACTTCC</td>\n",
       "      <td>GTCCCGAAGACTTCC</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>NUBPL</td>\n",
       "      <td>HGNC:20278</td>\n",
       "      <td>not provided|Inborn genetic diseases|Mitochond...</td>\n",
       "      <td>200401432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TATACGTGCCAGGTG</td>\n",
       "      <td>TATACGTACCAGGTG</td>\n",
       "      <td>Conflicting interpretations of pathogenicity; ...</td>\n",
       "      <td>germline</td>\n",
       "      <td>HFE</td>\n",
       "      <td>HGNC:4886</td>\n",
       "      <td>Hemochromatosis type 1|Porphyria cutanea tarda...</td>\n",
       "      <td>1800562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300086</th>\n",
       "      <td>GAACCGGGACCTTGC</td>\n",
       "      <td>GAACCGGCACCTTGC</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>GAA</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Glycogen storage disease, type II</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300087</th>\n",
       "      <td>CAGCAAGAGATGAGC</td>\n",
       "      <td>CAGCAAGTGATGAGC</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>DHTKD1</td>\n",
       "      <td>HGNC:23537</td>\n",
       "      <td>Charcot-Marie-Tooth disease axonal type 2Q</td>\n",
       "      <td>778723208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300088</th>\n",
       "      <td>GATGGGCCCAACGCC</td>\n",
       "      <td>GATGGGCTCAACGCC</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>COL2A1</td>\n",
       "      <td>HGNC:2200</td>\n",
       "      <td>Spondyloepiphyseal dysplasia congenita</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300089</th>\n",
       "      <td>CTATGCCAACAAGCA</td>\n",
       "      <td>CTATGCCGACAAGCA</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>HGNC:7632</td>\n",
       "      <td>Mucopolysaccharidosis, MPS-III-B</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300090</th>\n",
       "      <td>GGGCCGCGAGAGGCA</td>\n",
       "      <td>GGGCCGCTAGAGGCA</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>NOTCH3</td>\n",
       "      <td>HGNC:7883</td>\n",
       "      <td>Cerebral arteriopathy, autosomal dominant, wit...</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1300091 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ReferenceSequence  MutatedSequence  \\\n",
       "0         CACCTGCGGGTGAGT  CACCTGCAGGTGAGT   \n",
       "1         AAAGGTCCAGTCCTT  AAAGGTCTAGTCCTT   \n",
       "2         GTTGTCAACATGTAC  GTTGTCAGCATGTAC   \n",
       "3         GTCCCGAGGACTTCC  GTCCCGAAGACTTCC   \n",
       "4         TATACGTGCCAGGTG  TATACGTACCAGGTG   \n",
       "...                   ...              ...   \n",
       "1300086   GAACCGGGACCTTGC  GAACCGGCACCTTGC   \n",
       "1300087   CAGCAAGAGATGAGC  CAGCAAGTGATGAGC   \n",
       "1300088   GATGGGCCCAACGCC  GATGGGCTCAACGCC   \n",
       "1300089   CTATGCCAACAAGCA  CTATGCCGACAAGCA   \n",
       "1300090   GGGCCGCGAGAGGCA  GGGCCGCTAGAGGCA   \n",
       "\n",
       "                                      ClinicalSignificance OriginSimple  \\\n",
       "0                                   Uncertain significance     germline   \n",
       "1                                               Pathogenic     germline   \n",
       "2                                               Pathogenic     germline   \n",
       "3                                   Uncertain significance     germline   \n",
       "4        Conflicting interpretations of pathogenicity; ...     germline   \n",
       "...                                                    ...          ...   \n",
       "1300086                                  Likely pathogenic     germline   \n",
       "1300087                             Uncertain significance     germline   \n",
       "1300088                                  Likely pathogenic     germline   \n",
       "1300089                                  Likely pathogenic     germline   \n",
       "1300090                                  Likely pathogenic     germline   \n",
       "\n",
       "        GeneSymbol     HGNC_ID  \\\n",
       "0           ZNF592  HGNC:28986   \n",
       "1          FOXRED1  HGNC:26927   \n",
       "2          FOXRED1  HGNC:26927   \n",
       "3            NUBPL  HGNC:20278   \n",
       "4              HFE   HGNC:4886   \n",
       "...            ...         ...   \n",
       "1300086        GAA   HGNC:4065   \n",
       "1300087     DHTKD1  HGNC:23537   \n",
       "1300088     COL2A1   HGNC:2200   \n",
       "1300089      NAGLU   HGNC:7632   \n",
       "1300090     NOTCH3   HGNC:7883   \n",
       "\n",
       "                                             PhenotypeList  RS# (dbSNP)  \n",
       "0                                Galloway-Mowat syndrome 1    150829393  \n",
       "1        Mitochondrial complex 1 deficiency, nuclear ty...    267606829  \n",
       "2        Mitochondrial complex 1 deficiency, nuclear ty...    267606830  \n",
       "3        not provided|Inborn genetic diseases|Mitochond...    200401432  \n",
       "4        Hemochromatosis type 1|Porphyria cutanea tarda...      1800562  \n",
       "...                                                    ...          ...  \n",
       "1300086                  Glycogen storage disease, type II           -1  \n",
       "1300087         Charcot-Marie-Tooth disease axonal type 2Q    778723208  \n",
       "1300088             Spondyloepiphyseal dysplasia congenita           -1  \n",
       "1300089                   Mucopolysaccharidosis, MPS-III-B           -1  \n",
       "1300090  Cerebral arteriopathy, autosomal dominant, wit...           -1  \n",
       "\n",
       "[1300091 rows x 8 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hg19_processed_data = pd.DataFrame(hg19_processed_data)\n",
    "hg19_processed_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4fa86c9-32d4-4571-96ed-9c999ccc4c09",
   "metadata": {},
   "source": [
    "# Process GRCh38 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a10a92e2-e259-46a5-b9fe-9d5ceb6379da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "odict_keys(['NC_000001.11', 'NT_187361.1', 'NT_187362.1', 'NT_187363.1', 'NT_187364.1', 'NT_187365.1', 'NT_187366.1', 'NT_187367.1', 'NT_187368.1', 'NT_187369.1', 'NC_000002.12', 'NT_187370.1', 'NT_187371.1', 'NC_000003.12', 'NT_167215.1', 'NC_000004.12', 'NT_113793.3', 'NC_000005.10', 'NT_113948.1', 'NC_000006.12', 'NC_000007.14', 'NC_000008.11', 'NC_000009.12', 'NT_187372.1', 'NT_187373.1', 'NT_187374.1', 'NT_187375.1', 'NC_000010.11', 'NC_000011.10', 'NC_000012.12', 'NC_000013.11', 'NC_000014.9', 'NT_113796.3', 'NT_167219.1', 'NT_187377.1', 'NT_113888.1', 'NT_187378.1', 'NT_187379.1', 'NT_187380.1', 'NT_187381.1', 'NC_000015.10', 'NT_187382.1', 'NC_000016.10', 'NT_187383.1', 'NC_000017.11', 'NT_113930.2', 'NT_187384.1', 'NT_187385.1', 'NC_000018.10', 'NC_000019.10', 'NC_000020.11', 'NC_000021.9', 'NC_000022.11', 'NT_187386.1', 'NT_187387.1', 'NT_187388.1', 'NT_187390.1', 'NT_187391.1', 'NT_187392.1', 'NT_187393.1', 'NT_187394.1', 'NC_000023.11', 'NC_000024.10', 'NT_187395.1', 'NT_187396.1', 'NT_187397.1', 'NT_187398.1', 'NT_187399.1', 'NT_187400.1', 'NT_187401.1', 'NT_187402.1', 'NT_187403.1', 'NT_187404.1', 'NT_187405.1', 'NT_187406.1', 'NT_187407.1', 'NT_187408.1', 'NT_187409.1', 'NT_187410.1', 'NT_187411.1', 'NT_187412.1', 'NT_187413.1', 'NT_187414.1', 'NT_187415.1', 'NT_187416.1', 'NT_187417.1', 'NT_187418.1', 'NT_187419.1', 'NT_187420.1', 'NT_187421.1', 'NT_187422.1', 'NT_187423.1', 'NT_187424.1', 'NT_187425.1', 'NT_187426.1', 'NT_187427.1', 'NT_187428.1', 'NT_187429.1', 'NT_187430.1', 'NT_187431.1', 'NT_187432.1', 'NT_187433.1', 'NT_187434.1', 'NT_187435.1', 'NT_187436.1', 'NT_187437.1', 'NT_187438.1', 'NT_187439.1', 'NT_187440.1', 'NT_187441.1', 'NT_187442.1', 'NT_187443.1', 'NT_187444.1', 'NT_187445.1', 'NT_187446.1', 'NT_187447.1', 'NT_187448.1', 'NT_187449.1', 'NT_187450.1', 'NT_187451.1', 'NT_187452.1', 'NT_187453.1', 'NT_187454.1', 'NT_187455.1', 'NT_187456.1', 'NT_187457.1', 'NT_187458.1', 'NT_187459.1', 'NT_187460.1', 'NT_187461.1', 'NT_187462.1', 'NT_187463.1', 'NT_187464.1', 'NT_187465.1', 'NT_187466.1', 'NT_187467.1', 'NT_187468.1', 'NT_187469.1', 'NT_187470.1', 'NT_187471.1', 'NT_187472.1', 'NT_187473.1', 'NT_187474.1', 'NT_187475.1', 'NT_187476.1', 'NT_187477.1', 'NT_187478.1', 'NT_187479.1', 'NT_187480.1', 'NT_187481.1', 'NT_187482.1', 'NT_187483.1', 'NT_187484.1', 'NT_187485.1', 'NT_187486.1', 'NT_187487.1', 'NT_187488.1', 'NT_187489.1', 'NT_187490.1', 'NT_187491.1', 'NT_187492.1', 'NT_187493.1', 'NT_187494.1', 'NT_187495.1', 'NT_187496.1', 'NT_113901.1', 'NT_167213.1', 'NT_167214.1', 'NT_167218.1', 'NT_187497.1', 'NT_167220.1', 'NT_167208.1', 'NT_187498.1', 'NT_187499.1', 'NT_187500.1', 'NT_187501.1', 'NT_187502.1', 'NT_187503.1', 'NT_187504.1', 'NT_187505.1', 'NT_187506.1', 'NT_187508.1', 'NT_187509.1', 'NT_187510.1', 'NT_187511.1', 'NT_187512.1', 'NT_167209.1', 'NT_187513.1', 'NT_167211.2', 'NT_113889.1', 'NW_012132914.1', 'NW_015495298.1', 'NW_025791756.1', 'NW_011332688.1', 'NW_014040926.1', 'NW_009646195.1', 'NW_018654706.1', 'NW_019805487.1', 'NW_009646194.1', 'NW_018654707.1', 'NW_014040925.1', 'NW_017852928.1', 'NW_009646196.1', 'NW_025791753.1', 'NW_025791758.1', 'NW_025791759.1', 'NW_025791754.1', 'NW_011332687.1', 'NW_018654708.1', 'NW_014040927.1', 'NW_025791757.1', 'NW_025791755.1', 'NW_021159988.1', 'NW_025791767.1', 'NW_025791768.1', 'NW_025791766.1', 'NW_025791763.1', 'NW_012132915.1', 'NW_025791760.1', 'NW_025791765.1', 'NW_018654709.1', 'NW_025791762.1', 'NW_025791761.1', 'NW_025791764.1', 'NW_015495299.1', 'NW_018654710.1', 'NW_011332690.1', 'NW_021159987.1', 'NW_011332689.1', 'NW_017363813.1', 'NW_025791770.1', 'NW_025791771.1', 'NW_009646197.1', 'NW_012132916.1', 'NW_011332691.1', 'NW_021159989.1', 'NW_018654711.1', 'NW_012132917.1', 'NW_009646198.1', 'NW_019805491.1', 'NW_019805492.1', 'NW_019805490.1', 'NW_019805489.1', 'NW_019805488.1', 'NW_025791769.1', 'NW_021159990.1', 'NW_021159993.1', 'NW_013171799.1', 'NW_025791774.1', 'NW_021159992.1', 'NW_021159994.1', 'NW_021159991.1', 'NW_021159995.1', 'NW_013171800.1', 'NW_013171801.1', 'NW_025791772.1', 'NW_017363814.1', 'NW_025791773.1', 'NW_015495301.1', 'NW_015495300.1', 'NW_025791777.1', 'NW_018654712.1', 'NW_025791776.1', 'NW_009646199.1', 'NW_016107297.1', 'NW_025791779.1', 'NW_021159996.1', 'NW_025791778.1', 'NW_016107298.1', 'NW_025791775.1', 'NW_018654713.1', 'NW_013171803.1', 'NW_025791780.1', 'NW_012132918.1', 'NW_009646200.1', 'NW_013171802.1', 'NW_017363815.1', 'NW_021159997.1', 'NW_021159998.1', 'NW_019805493.1', 'NW_025791781.1', 'NW_017852929.1', 'NW_017852930.1', 'NW_018654714.1', 'NW_018654715.1', 'NW_012132919.1', 'NW_025791785.1', 'NW_018654717.1', 'NW_017852932.1', 'NW_025791782.1', 'NW_017852931.1', 'NW_025791784.1', 'NW_019805494.1', 'NW_025791786.1', 'NW_018654716.1', 'NW_025791783.1', 'NW_013171804.1', 'NW_013171805.1', 'NW_025791789.1', 'NW_025791787.1', 'NW_025791788.1', 'NW_009646201.1', 'NW_021159999.1', 'NW_021160000.1', 'NW_011332694.1', 'NW_021160001.1', 'NW_013171806.1', 'NW_009646202.1', 'NW_013171807.1', 'NW_025791790.1', 'NW_011332693.1', 'NW_011332692.1', 'NW_015148966.2', 'NW_025791792.1', 'NW_021160004.1', 'NW_025791794.1', 'NW_011332695.1', 'NW_021160006.1', 'NW_019805496.1', 'NW_019805495.1', 'NW_017363816.1', 'NW_025791793.1', 'NW_019805498.1', 'NW_021160005.1', 'NW_021160003.1', 'NW_019805497.1', 'NW_013171808.1', 'NW_021160002.1', 'NW_025791791.1', 'NW_009646203.1', 'NW_013171809.1', 'NW_018654718.1', 'NW_021160008.1', 'NW_011332696.1', 'NW_009646204.1', 'NW_025791795.1', 'NW_018654720.1', 'NW_015148967.1', 'NW_018654719.1', 'NW_011332697.1', 'NW_019805499.1', 'NW_021160007.1', 'NW_021160012.1', 'NW_011332699.1', 'NW_013171810.1', 'NW_021160009.1', 'NW_009646205.1', 'NW_011332700.1', 'NW_013171811.1', 'NW_021160010.1', 'NW_021160011.1', 'NW_011332698.1', 'NW_021160013.1', 'NW_025791796.1', 'NW_018654722.1', 'NW_021160014.1', 'NW_018654721.1', 'NW_011332701.1', 'NW_021160018.1', 'NW_021160017.1', 'NW_012132920.1', 'NW_025791798.1', 'NW_021160016.1', 'NW_025791797.1', 'NW_021160015.1', 'NW_025791799.1', 'NW_013171812.1', 'NW_019805500.1', 'NW_017852933.1', 'NW_021160019.1', 'NW_013171813.1', 'NW_018654723.1', 'NW_012132921.1', 'NW_025791800.1', 'NW_017363817.1', 'NW_021160020.1', 'NW_016107299.1', 'NW_017363819.1', 'NW_025791803.1', 'NW_025791801.1', 'NW_017363818.1', 'NW_019805501.1', 'NW_025791806.1', 'NW_025791802.1', 'NW_025791805.1', 'NW_021160021.1', 'NW_025791804.1', 'NW_019805503.1', 'NW_014040928.1', 'NW_019805502.1', 'NW_013171814.1', 'NW_018654724.1', 'NW_025791809.1', 'NW_025791810.1', 'NW_025791807.1', 'NW_021160022.1', 'NW_014040929.1', 'NW_025791808.1', 'NW_009646206.1', 'NW_016107300.1', 'NW_016107301.1', 'NW_016107302.1', 'NW_016107303.1', 'NW_016107304.1', 'NW_016107305.1', 'NW_016107306.1', 'NW_016107307.1', 'NW_016107308.1', 'NW_016107309.1', 'NW_016107310.1', 'NW_016107311.1', 'NW_016107313.1', 'NW_016107314.1', 'NW_016107312.1', 'NW_025791811.1', 'NW_025791812.1', 'NW_021160023.1', 'NW_025791813.1', 'NW_025791814.1', 'NW_025791815.1', 'NW_021160026.1', 'NW_021160024.1', 'NW_009646207.1', 'NW_014040930.1', 'NW_014040931.1', 'NW_009646208.1', 'NW_015148968.1', 'NW_021160025.1', 'NW_015148969.2', 'NW_017363820.1', 'NW_021160031.1', 'NW_025791820.1', 'NW_021160028.1', 'NW_025791816.1', 'NW_021160029.1', 'NW_025791817.1', 'NW_021160027.1', 'NW_025791819.1', 'NW_021160030.1', 'NW_025791818.1', 'NW_018654725.1', 'NW_025791821.1', 'NW_018654726.1', 'NW_009646209.1', 'NT_187515.1', 'NT_187517.1', 'NT_187514.1', 'NT_187520.1', 'NW_003315905.1', 'NW_003315906.1', 'NW_003315907.2', 'NT_187521.1', 'NT_187519.1', 'NT_187516.1', 'NT_187518.1', 'NT_187525.1', 'NT_187526.1', 'NT_187529.1', 'NT_187522.1', 'NW_003315908.1', 'NT_187524.1', 'NT_187531.1', 'NT_187530.1', 'NT_187528.1', 'NW_003571033.2', 'NW_003315909.1', 'NT_187527.1', 'NT_187523.1', 'NW_003871060.2', 'NT_187535.1', 'NT_187537.1', 'NW_003315913.1', 'NT_187533.1', 'NT_187536.1', 'NT_187538.1', 'NT_187532.1', 'NT_187534.1', 'NT_187539.1', 'NT_187540.1', 'NW_003315915.1', 'NT_187541.1', 'NT_167250.2', 'NT_187544.1', 'NW_003315914.1', 'NT_187542.1', 'NT_187545.1', 'NT_187543.1', 'NT_187550.1', 'NT_187548.1', 'NT_187547.1', 'NW_003315920.1', 'NW_003571036.1', 'NT_187551.1', 'NW_003315917.2', 'NW_003315918.1', 'NT_187549.1', 'NW_003315919.1', 'NT_187546.1', 'NT_167244.2', 'NT_187555.1', 'NT_187554.1', 'NW_003315921.1', 'NT_187556.1', 'NT_187557.1', 'NW_004166862.2', 'NT_187552.1', 'NT_187553.1', 'NT_187558.1', 'NT_187561.1', 'NT_187559.1', 'NW_003315922.2', 'NT_187562.1', 'NT_187564.1', 'NT_187563.1', 'NT_187560.1', 'NT_187572.1', 'NT_187568.1', 'NT_187565.1', 'NT_187576.1', 'NT_187570.1', 'NT_187577.1', 'NT_187566.1', 'NT_187567.1', 'NT_187574.1', 'NT_187575.1', 'NT_187573.1', 'NT_187571.1', 'NT_187569.1', 'NW_003315928.1', 'NW_003315929.1', 'NW_003315930.1', 'NW_003315931.1', 'NT_187578.1', 'NW_003315934.1', 'NT_187579.1', 'NW_003315935.1', 'NT_187586.1', 'NT_187584.1', 'NT_187585.1', 'NT_187583.1', 'NW_003315936.1', 'NW_003871073.1', 'NW_003871074.1', 'NT_187582.1', 'NT_187581.1', 'NW_003571049.1', 'NW_003571050.1', 'NT_187588.1', 'NW_003315938.1', 'NT_187587.1', 'NW_003315939.2', 'NW_003315941.1', 'NW_003315942.2', 'NT_187590.1', 'NW_003315940.1', 'NT_187589.1', 'NT_187591.1', 'NT_187594.1', 'NT_187593.1', 'NT_187597.1', 'NT_187595.1', 'NT_187592.1', 'NT_187596.1', 'NT_187598.1', 'NT_187601.1', 'NT_187599.1', 'NT_187600.1', 'NT_187602.1', 'NT_187604.1', 'NT_187603.1', 'NW_003315943.1', 'NT_187605.1', 'NW_003315944.2', 'NT_187606.1', 'NT_187610.1', 'NT_187609.1', 'NT_187608.1', 'NT_187607.1', 'NW_003315945.1', 'NW_003315946.1', 'NW_003315952.3', 'NT_187613.1', 'NT_187611.1', 'NT_187614.1', 'NW_003871091.1', 'NW_003871092.1', 'NW_003315953.2', 'NT_167251.2', 'NW_003315954.1', 'NT_187615.1', 'NT_187616.1', 'NW_003315955.1', 'NT_187612.1', 'NT_187618.1', 'NW_003315956.1', 'NW_003315959.1', 'NW_003315960.1', 'NW_003315957.1', 'NW_003315958.1', 'NW_003315961.1', 'NT_187617.1', 'NT_187622.1', 'NT_187621.1', 'NW_003315962.1', 'NW_003315964.2', 'NW_003315965.1', 'NW_003315963.1', 'NT_187619.1', 'NT_187620.1', 'NW_003571054.1', 'NW_003315966.2', 'NT_187623.1', 'NT_187625.1', 'NT_187624.1', 'NW_003315967.2', 'NT_187628.1', 'NT_187627.1', 'NW_003315968.2', 'NW_003315969.2', 'NW_003315970.2', 'NT_187626.1', 'NT_187629.1', 'NT_187632.1', 'NT_187633.1', 'NT_187630.1', 'NT_187631.1', 'NW_003315972.2', 'NW_003315971.2', 'NT_187634.1', 'NT_187635.1', 'NT_187646.1', 'NT_187648.1', 'NT_187647.1', 'NT_187649.1', 'NT_187650.1', 'NT_187651.1', 'NT_187652.1', 'NT_113891.3', 'NT_187653.1', 'NT_187655.1', 'NT_187654.1', 'NT_187656.1', 'NT_187657.1', 'NT_187658.1', 'NT_187659.1', 'NT_187660.1', 'NT_187662.1', 'NT_187664.1', 'NT_187661.1', 'NW_003871093.1', 'NT_187663.1', 'NT_187665.1', 'NT_187666.1', 'NW_003571055.2', 'NW_004504305.1', 'NT_187667.1', 'NT_187678.1', 'NT_187679.1', 'NT_167245.2', 'NT_187680.1', 'NT_187681.1', 'NW_003571056.2', 'NT_187682.1', 'NT_187688.1', 'NT_167246.2', 'NW_003571057.2', 'NT_187689.1', 'NT_167247.2', 'NW_003571058.2', 'NT_187690.1', 'NT_167248.2', 'NW_003571059.2', 'NT_187691.1', 'NT_167249.2', 'NW_003571060.1', 'NT_187692.1', 'NW_003571061.2', 'NT_187693.1', 'NT_187636.1', 'NT_187637.1', 'NT_187638.1', 'NT_187639.1', 'NT_187640.1', 'NT_187641.1', 'NT_187642.1', 'NT_187643.1', 'NT_187644.1', 'NT_187645.1', 'NT_187668.1', 'NT_187669.1', 'NT_187670.1', 'NT_187671.1', 'NT_187672.1', 'NT_187673.1', 'NT_187674.1', 'NT_187675.1', 'NT_187676.1', 'NT_187677.1', 'NT_187683.1', 'NT_187684.1', 'NT_187685.1', 'NT_187686.1', 'NT_187687.1', 'NT_113949.2', 'NC_012920.1'])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch38_genome = Fasta('data/reference_genomes/GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.fna')\n",
    "grch38_genome.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "15d80eb3-c520-4611-80c1-675e5d4c7dc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1': 'NC_000001.11',\n",
       " '2': 'NC_000002.12',\n",
       " '3': 'NC_000003.12',\n",
       " '4': 'NC_000004.12',\n",
       " '5': 'NC_000005.10',\n",
       " '6': 'NC_000006.12',\n",
       " '7': 'NC_000007.14',\n",
       " '8': 'NC_000008.11',\n",
       " '9': 'NC_000009.12',\n",
       " '10': 'NC_000010.11',\n",
       " '11': 'NC_000011.10',\n",
       " '12': 'NC_000012.12',\n",
       " '13': 'NC_000013.11',\n",
       " '14': 'NC_000014.9',\n",
       " '15': 'NC_000015.10',\n",
       " '16': 'NC_000016.10',\n",
       " '17': 'NC_000017.11',\n",
       " '18': 'NC_000018.10',\n",
       " '19': 'NC_000019.10',\n",
       " '20': 'NC_000020.11',\n",
       " '21': 'NC_000021.9',\n",
       " '22': 'NC_000022.11',\n",
       " 'X': 'NC_000023.11',\n",
       " 'Y': 'NC_000024.10'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch38_chrom_mapping = {}\n",
    "for chrom in valid_chromosomes:\n",
    "    c = chrom\n",
    "    if chrom==\"X\":\n",
    "        c = \"23\"\n",
    "    elif chrom==\"Y\":\n",
    "        c = \"24\"\n",
    "    ref = [k for k in grch38_genome.keys() if \"NC_00\" in k and \"0\"+c+\".\" in k][-1]\n",
    "    grch38_chrom_mapping[chrom] = ref\n",
    "\n",
    "grch38_chrom_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f381d477-bc9c-4496-83d3-dbb858c28b48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>150829393</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>4</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>84799209</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15044</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.694C&gt;T (p.Gln232Ter)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>267606829</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>5</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>126275389</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15045</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_017547.4(FOXRED1):c.1289A&gt;G (p.Asn430Ser)</td>\n",
       "      <td>55572</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>267606830</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>6</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>126277517</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>15046</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_025152.3(NUBPL):c.166G&gt;A (p.Gly56Arg)</td>\n",
       "      <td>80224</td>\n",
       "      <td>NUBPL</td>\n",
       "      <td>HGNC:20278</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>200401432</td>\n",
       "      <td>not provided|Inborn genetic diseases|Mitochond...</td>\n",
       "      <td>germline;paternal</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>214885</td>\n",
       "      <td>NC_000014.9</td>\n",
       "      <td>14</td>\n",
       "      <td>31562125</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>15048</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000410.4(HFE):c.845G&gt;A (p.Cys282Tyr)</td>\n",
       "      <td>3077</td>\n",
       "      <td>HFE</td>\n",
       "      <td>HGNC:4886</td>\n",
       "      <td>Conflicting interpretations of pathogenicity; ...</td>\n",
       "      <td>1800562</td>\n",
       "      <td>Hemochromatosis type 1|Porphyria cutanea tarda...</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>9</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>6</td>\n",
       "      <td>26092913</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835682</th>\n",
       "      <td>1952235</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000152.5(GAA):c.844G&gt;C (p.Asp282His)</td>\n",
       "      <td>2548</td>\n",
       "      <td>GAA</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Glycogen storage disease, type II</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895435</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>80107708</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835684</th>\n",
       "      <td>1952237</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_018706.7(DHTKD1):c.2546A&gt;T (p.Glu849Val)</td>\n",
       "      <td>55526</td>\n",
       "      <td>DHTKD1</td>\n",
       "      <td>HGNC:23537</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>778723208</td>\n",
       "      <td>Charcot-Marie-Tooth disease axonal type 2Q</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895437</td>\n",
       "      <td>NC_000010.11</td>\n",
       "      <td>10</td>\n",
       "      <td>12118892</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835686</th>\n",
       "      <td>1952238</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001844.5(COL2A1):c.1502G&gt;A (p.Gly501Glu)</td>\n",
       "      <td>1280</td>\n",
       "      <td>COL2A1</td>\n",
       "      <td>HGNC:2200</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Spondyloepiphyseal dysplasia congenita</td>\n",
       "      <td>de novo</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895438</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>47986361</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835688</th>\n",
       "      <td>1952239</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000263.4(NAGLU):c.1978A&gt;G (p.Asn660Asp)</td>\n",
       "      <td>4669</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>HGNC:7632</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Mucopolysaccharidosis, MPS-III-B</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895439</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>42543984</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2835690</th>\n",
       "      <td>1952240</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000435.3(NOTCH3):c.2903C&gt;A (p.Ser968Ter)</td>\n",
       "      <td>4854</td>\n",
       "      <td>NOTCH3</td>\n",
       "      <td>HGNC:7883</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>-1</td>\n",
       "      <td>Cerebral arteriopathy, autosomal dominant, wit...</td>\n",
       "      <td>paternal</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>1895440</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>15181052</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1417597 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #AlleleID                       Type  \\\n",
       "1            15043  single nucleotide variant   \n",
       "3            15044  single nucleotide variant   \n",
       "5            15045  single nucleotide variant   \n",
       "7            15046  single nucleotide variant   \n",
       "9            15048  single nucleotide variant   \n",
       "...            ...                        ...   \n",
       "2835682    1952235  single nucleotide variant   \n",
       "2835684    1952237  single nucleotide variant   \n",
       "2835686    1952238  single nucleotide variant   \n",
       "2835688    1952239  single nucleotide variant   \n",
       "2835690    1952240  single nucleotide variant   \n",
       "\n",
       "                                                 Name  GeneID GeneSymbol  \\\n",
       "1        NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592   \n",
       "3         NM_017547.4(FOXRED1):c.694C>T (p.Gln232Ter)   55572    FOXRED1   \n",
       "5        NM_017547.4(FOXRED1):c.1289A>G (p.Asn430Ser)   55572    FOXRED1   \n",
       "7            NM_025152.3(NUBPL):c.166G>A (p.Gly56Arg)   80224      NUBPL   \n",
       "9             NM_000410.4(HFE):c.845G>A (p.Cys282Tyr)    3077        HFE   \n",
       "...                                               ...     ...        ...   \n",
       "2835682       NM_000152.5(GAA):c.844G>C (p.Asp282His)    2548        GAA   \n",
       "2835684   NM_018706.7(DHTKD1):c.2546A>T (p.Glu849Val)   55526     DHTKD1   \n",
       "2835686   NM_001844.5(COL2A1):c.1502G>A (p.Gly501Glu)    1280     COL2A1   \n",
       "2835688    NM_000263.4(NAGLU):c.1978A>G (p.Asn660Asp)    4669      NAGLU   \n",
       "2835690   NM_000435.3(NOTCH3):c.2903C>A (p.Ser968Ter)    4854     NOTCH3   \n",
       "\n",
       "            HGNC_ID                               ClinicalSignificance  \\\n",
       "1        HGNC:28986                             Uncertain significance   \n",
       "3        HGNC:26927                                         Pathogenic   \n",
       "5        HGNC:26927                                         Pathogenic   \n",
       "7        HGNC:20278                             Uncertain significance   \n",
       "9         HGNC:4886  Conflicting interpretations of pathogenicity; ...   \n",
       "...             ...                                                ...   \n",
       "2835682   HGNC:4065                                  Likely pathogenic   \n",
       "2835684  HGNC:23537                             Uncertain significance   \n",
       "2835686   HGNC:2200                                  Likely pathogenic   \n",
       "2835688   HGNC:7632                                  Likely pathogenic   \n",
       "2835690   HGNC:7883                                  Likely pathogenic   \n",
       "\n",
       "         RS# (dbSNP)                                      PhenotypeList  \\\n",
       "1          150829393                          Galloway-Mowat syndrome 1   \n",
       "3          267606829  Mitochondrial complex 1 deficiency, nuclear ty...   \n",
       "5          267606830  Mitochondrial complex 1 deficiency, nuclear ty...   \n",
       "7          200401432  not provided|Inborn genetic diseases|Mitochond...   \n",
       "9            1800562  Hemochromatosis type 1|Porphyria cutanea tarda...   \n",
       "...              ...                                                ...   \n",
       "2835682           -1                  Glycogen storage disease, type II   \n",
       "2835684    778723208         Charcot-Marie-Tooth disease axonal type 2Q   \n",
       "2835686           -1             Spondyloepiphyseal dysplasia congenita   \n",
       "2835688           -1                   Mucopolysaccharidosis, MPS-III-B   \n",
       "2835690           -1  Cerebral arteriopathy, autosomal dominant, wit...   \n",
       "\n",
       "                    Origin OriginSimple Assembly  VariationID  \\\n",
       "1                 germline     germline   GRCh38            4   \n",
       "3         germline;unknown     germline   GRCh38            5   \n",
       "5                 germline     germline   GRCh38            6   \n",
       "7        germline;paternal     germline   GRCh38       214885   \n",
       "9         germline;unknown     germline   GRCh38            9   \n",
       "...                    ...          ...      ...          ...   \n",
       "2835682           germline     germline   GRCh38      1895435   \n",
       "2835684           germline     germline   GRCh38      1895437   \n",
       "2835686            de novo     germline   GRCh38      1895438   \n",
       "2835688           germline     germline   GRCh38      1895439   \n",
       "2835690           paternal     germline   GRCh38      1895440   \n",
       "\n",
       "        ChromosomeAccession Chromosome  PositionVCF ReferenceAlleleVCF  \\\n",
       "1              NC_000015.10         15     84799209                  G   \n",
       "3              NC_000011.10         11    126275389                  C   \n",
       "5              NC_000011.10         11    126277517                  A   \n",
       "7               NC_000014.9         14     31562125                  G   \n",
       "9              NC_000006.12          6     26092913                  G   \n",
       "...                     ...        ...          ...                ...   \n",
       "2835682        NC_000017.11         17     80107708                  G   \n",
       "2835684        NC_000010.11         10     12118892                  A   \n",
       "2835686        NC_000012.12         12     47986361                  C   \n",
       "2835688        NC_000017.11         17     42543984                  A   \n",
       "2835690        NC_000019.10         19     15181052                  G   \n",
       "\n",
       "        AlternateAlleleVCF  \n",
       "1                        A  \n",
       "3                        T  \n",
       "5                        G  \n",
       "7                        A  \n",
       "9                        A  \n",
       "...                    ...  \n",
       "2835682                  C  \n",
       "2835684                  T  \n",
       "2835686                  T  \n",
       "2835688                  G  \n",
       "2835690                  T  \n",
       "\n",
       "[1417597 rows x 18 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch38_data = data[data[\"Assembly\"]==\"GRCh38\"]\n",
    "grch38_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "5fbb78f2-d9ac-4c20-b26e-703c8ea04cd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "122d8dc2d0b1457089a4911427b7c8f8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "0it [00:00, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "grch38_processed_data = []\n",
    "for i, row in tqdm(grch38_data.iterrows()):\n",
    "    if row[\"Chromosome\"] not in valid_chromosomes:\n",
    "        continue\n",
    "        \n",
    "        \n",
    "    chrom = grch38_chrom_mapping[row[\"Chromosome\"]]\n",
    "    ix = row[\"PositionVCF\"]-1\n",
    "    if not grch38_genome[chrom][ix]==row.ReferenceAlleleVCF:\n",
    "        continue\n",
    "\n",
    "    preceding_sequence = grch38_genome[chrom][ix-buffer_size:ix]\n",
    "    following_sequence = grch38_genome[chrom][ix+1:ix+buffer_size+1]\n",
    "    grch38_processed_data.append({\"ReferenceSequence\": str(preceding_sequence).upper()+row.ReferenceAlleleVCF+str(following_sequence).upper(), \n",
    "                           \"MutatedSequence\": str(preceding_sequence).upper()+row.AlternateAlleleVCF+str(following_sequence).upper(),\n",
    "                          \"ClinicalSignificance\": row.ClinicalSignificance, \"OriginSimple\": row.OriginSimple, \"GeneSymbol\": row.GeneSymbol, \"HGNC_ID\": row.HGNC_ID, \"PhenotypeList\": row.PhenotypeList,\n",
    "                           \"RS# (dbSNP)\": row[\"RS# (dbSNP)\"]\n",
    "                          })\n",
    "    #if i>30:\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "23cf55f7-1505-45cc-b468-c3f92c6d12f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ReferenceSequence</th>\n",
       "      <th>MutatedSequence</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CACCTGCGGGTGAGT</td>\n",
       "      <td>CACCTGCAGGTGAGT</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>150829393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGTCCAGTCCTT</td>\n",
       "      <td>AAAGGTCTAGTCCTT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>267606829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTTGTCAACATGTAC</td>\n",
       "      <td>GTTGTCAGCATGTAC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>267606830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTCCCGAGGACTTCC</td>\n",
       "      <td>GTCCCGAAGACTTCC</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>NUBPL</td>\n",
       "      <td>HGNC:20278</td>\n",
       "      <td>not provided|Inborn genetic diseases|Mitochond...</td>\n",
       "      <td>200401432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TATACGTGCCAGGTG</td>\n",
       "      <td>TATACGTACCAGGTG</td>\n",
       "      <td>Conflicting interpretations of pathogenicity; ...</td>\n",
       "      <td>germline</td>\n",
       "      <td>HFE</td>\n",
       "      <td>HGNC:4886</td>\n",
       "      <td>Hemochromatosis type 1|Porphyria cutanea tarda...</td>\n",
       "      <td>1800562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207701</th>\n",
       "      <td>GAACCGGGACCTTGC</td>\n",
       "      <td>GAACCGGCACCTTGC</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>GAA</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Glycogen storage disease, type II</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207702</th>\n",
       "      <td>CAGCAAGAGATGAGC</td>\n",
       "      <td>CAGCAAGTGATGAGC</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>DHTKD1</td>\n",
       "      <td>HGNC:23537</td>\n",
       "      <td>Charcot-Marie-Tooth disease axonal type 2Q</td>\n",
       "      <td>778723208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207703</th>\n",
       "      <td>GATGGGCCCAACGCC</td>\n",
       "      <td>GATGGGCTCAACGCC</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>COL2A1</td>\n",
       "      <td>HGNC:2200</td>\n",
       "      <td>Spondyloepiphyseal dysplasia congenita</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207704</th>\n",
       "      <td>CTATGCCAACAAGCA</td>\n",
       "      <td>CTATGCCGACAAGCA</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>HGNC:7632</td>\n",
       "      <td>Mucopolysaccharidosis, MPS-III-B</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207705</th>\n",
       "      <td>GGGCCGCGAGAGGCA</td>\n",
       "      <td>GGGCCGCTAGAGGCA</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>NOTCH3</td>\n",
       "      <td>HGNC:7883</td>\n",
       "      <td>Cerebral arteriopathy, autosomal dominant, wit...</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1207706 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ReferenceSequence  MutatedSequence  \\\n",
       "0         CACCTGCGGGTGAGT  CACCTGCAGGTGAGT   \n",
       "1         AAAGGTCCAGTCCTT  AAAGGTCTAGTCCTT   \n",
       "2         GTTGTCAACATGTAC  GTTGTCAGCATGTAC   \n",
       "3         GTCCCGAGGACTTCC  GTCCCGAAGACTTCC   \n",
       "4         TATACGTGCCAGGTG  TATACGTACCAGGTG   \n",
       "...                   ...              ...   \n",
       "1207701   GAACCGGGACCTTGC  GAACCGGCACCTTGC   \n",
       "1207702   CAGCAAGAGATGAGC  CAGCAAGTGATGAGC   \n",
       "1207703   GATGGGCCCAACGCC  GATGGGCTCAACGCC   \n",
       "1207704   CTATGCCAACAAGCA  CTATGCCGACAAGCA   \n",
       "1207705   GGGCCGCGAGAGGCA  GGGCCGCTAGAGGCA   \n",
       "\n",
       "                                      ClinicalSignificance OriginSimple  \\\n",
       "0                                   Uncertain significance     germline   \n",
       "1                                               Pathogenic     germline   \n",
       "2                                               Pathogenic     germline   \n",
       "3                                   Uncertain significance     germline   \n",
       "4        Conflicting interpretations of pathogenicity; ...     germline   \n",
       "...                                                    ...          ...   \n",
       "1207701                                  Likely pathogenic     germline   \n",
       "1207702                             Uncertain significance     germline   \n",
       "1207703                                  Likely pathogenic     germline   \n",
       "1207704                                  Likely pathogenic     germline   \n",
       "1207705                                  Likely pathogenic     germline   \n",
       "\n",
       "        GeneSymbol     HGNC_ID  \\\n",
       "0           ZNF592  HGNC:28986   \n",
       "1          FOXRED1  HGNC:26927   \n",
       "2          FOXRED1  HGNC:26927   \n",
       "3            NUBPL  HGNC:20278   \n",
       "4              HFE   HGNC:4886   \n",
       "...            ...         ...   \n",
       "1207701        GAA   HGNC:4065   \n",
       "1207702     DHTKD1  HGNC:23537   \n",
       "1207703     COL2A1   HGNC:2200   \n",
       "1207704      NAGLU   HGNC:7632   \n",
       "1207705     NOTCH3   HGNC:7883   \n",
       "\n",
       "                                             PhenotypeList  RS# (dbSNP)  \n",
       "0                                Galloway-Mowat syndrome 1    150829393  \n",
       "1        Mitochondrial complex 1 deficiency, nuclear ty...    267606829  \n",
       "2        Mitochondrial complex 1 deficiency, nuclear ty...    267606830  \n",
       "3        not provided|Inborn genetic diseases|Mitochond...    200401432  \n",
       "4        Hemochromatosis type 1|Porphyria cutanea tarda...      1800562  \n",
       "...                                                    ...          ...  \n",
       "1207701                  Glycogen storage disease, type II           -1  \n",
       "1207702         Charcot-Marie-Tooth disease axonal type 2Q    778723208  \n",
       "1207703             Spondyloepiphyseal dysplasia congenita           -1  \n",
       "1207704                   Mucopolysaccharidosis, MPS-III-B           -1  \n",
       "1207705  Cerebral arteriopathy, autosomal dominant, wit...           -1  \n",
       "\n",
       "[1207706 rows x 8 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grch38_processed_data = pd.DataFrame(grch38_processed_data)\n",
    "grch38_processed_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e11b9337-12b6-4841-be0b-e3f2dde11b83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "#AlleleID                                                         1952240\n",
       "Type                                            single nucleotide variant\n",
       "Name                          NM_000435.3(NOTCH3):c.2903C>A (p.Ser968Ter)\n",
       "GeneID                                                               4854\n",
       "GeneSymbol                                                         NOTCH3\n",
       "HGNC_ID                                                         HGNC:7883\n",
       "ClinicalSignificance                                    Likely pathogenic\n",
       "RS# (dbSNP)                                                            -1\n",
       "PhenotypeList           Cerebral arteriopathy, autosomal dominant, wit...\n",
       "Origin                                                           paternal\n",
       "OriginSimple                                                     germline\n",
       "Assembly                                                           GRCh38\n",
       "VariationID                                                       1895440\n",
       "ChromosomeAccession                                          NC_000019.10\n",
       "Chromosome                                                             19\n",
       "PositionVCF                                                      15181052\n",
       "ReferenceAlleleVCF                                                      G\n",
       "AlternateAlleleVCF                                                      T\n",
       "Name: 2835690, dtype: object"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "60427f36-8383-4b35-8763-dc72c0b78dc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_processed_data = pd.concat([hg19_processed_data, grch38_processed_data])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "88957cc4-4acf-442e-ab94-f4a512c4c36a",
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_processed_data.to_csv(\"processed_data/clinvar_processed_snvs.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "defa9967-3400-4bf5-a8d8-1d41f3607cd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Uncertain significance                          1033627\n",
       "Likely benign                                    783749\n",
       "Benign                                           275297\n",
       "Pathogenic                                       115338\n",
       "Conflicting interpretations of pathogenicity     114752\n",
       "                                                 ...   \n",
       "Benign/Likely benign; association                     2\n",
       "Benign; protective                                    2\n",
       "association; drug response; risk factor               2\n",
       "Pathogenic; protective                                2\n",
       "Likely benign; association                            1\n",
       "Name: ClinicalSignificance, Length: 81, dtype: int64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_processed_data[\"ClinicalSignificance\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "73b6511d-ec1f-4afb-b88a-1c053e12914b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ReferenceSequence</th>\n",
       "      <th>MutatedSequence</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CACCTGCGGGTGAGT</td>\n",
       "      <td>CACCTGCAGGTGAGT</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>150829393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAGGTCCAGTCCTT</td>\n",
       "      <td>AAAGGTCTAGTCCTT</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>267606829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTTGTCAACATGTAC</td>\n",
       "      <td>GTTGTCAGCATGTAC</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>HGNC:26927</td>\n",
       "      <td>Mitochondrial complex 1 deficiency, nuclear ty...</td>\n",
       "      <td>267606830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTCCCGAGGACTTCC</td>\n",
       "      <td>GTCCCGAAGACTTCC</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>NUBPL</td>\n",
       "      <td>HGNC:20278</td>\n",
       "      <td>not provided|Inborn genetic diseases|Mitochond...</td>\n",
       "      <td>200401432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TATACGTGCCAGGTG</td>\n",
       "      <td>TATACGTACCAGGTG</td>\n",
       "      <td>Conflicting interpretations of pathogenicity; ...</td>\n",
       "      <td>germline</td>\n",
       "      <td>HFE</td>\n",
       "      <td>HGNC:4886</td>\n",
       "      <td>Hemochromatosis type 1|Porphyria cutanea tarda...</td>\n",
       "      <td>1800562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207701</th>\n",
       "      <td>GAACCGGGACCTTGC</td>\n",
       "      <td>GAACCGGCACCTTGC</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>GAA</td>\n",
       "      <td>HGNC:4065</td>\n",
       "      <td>Glycogen storage disease, type II</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207702</th>\n",
       "      <td>CAGCAAGAGATGAGC</td>\n",
       "      <td>CAGCAAGTGATGAGC</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>germline</td>\n",
       "      <td>DHTKD1</td>\n",
       "      <td>HGNC:23537</td>\n",
       "      <td>Charcot-Marie-Tooth disease axonal type 2Q</td>\n",
       "      <td>778723208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207703</th>\n",
       "      <td>GATGGGCCCAACGCC</td>\n",
       "      <td>GATGGGCTCAACGCC</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>COL2A1</td>\n",
       "      <td>HGNC:2200</td>\n",
       "      <td>Spondyloepiphyseal dysplasia congenita</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207704</th>\n",
       "      <td>CTATGCCAACAAGCA</td>\n",
       "      <td>CTATGCCGACAAGCA</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>HGNC:7632</td>\n",
       "      <td>Mucopolysaccharidosis, MPS-III-B</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207705</th>\n",
       "      <td>GGGCCGCGAGAGGCA</td>\n",
       "      <td>GGGCCGCTAGAGGCA</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>germline</td>\n",
       "      <td>NOTCH3</td>\n",
       "      <td>HGNC:7883</td>\n",
       "      <td>Cerebral arteriopathy, autosomal dominant, wit...</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2507797 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ReferenceSequence  MutatedSequence  \\\n",
       "0         CACCTGCGGGTGAGT  CACCTGCAGGTGAGT   \n",
       "1         AAAGGTCCAGTCCTT  AAAGGTCTAGTCCTT   \n",
       "2         GTTGTCAACATGTAC  GTTGTCAGCATGTAC   \n",
       "3         GTCCCGAGGACTTCC  GTCCCGAAGACTTCC   \n",
       "4         TATACGTGCCAGGTG  TATACGTACCAGGTG   \n",
       "...                   ...              ...   \n",
       "1207701   GAACCGGGACCTTGC  GAACCGGCACCTTGC   \n",
       "1207702   CAGCAAGAGATGAGC  CAGCAAGTGATGAGC   \n",
       "1207703   GATGGGCCCAACGCC  GATGGGCTCAACGCC   \n",
       "1207704   CTATGCCAACAAGCA  CTATGCCGACAAGCA   \n",
       "1207705   GGGCCGCGAGAGGCA  GGGCCGCTAGAGGCA   \n",
       "\n",
       "                                      ClinicalSignificance OriginSimple  \\\n",
       "0                                   Uncertain significance     germline   \n",
       "1                                               Pathogenic     germline   \n",
       "2                                               Pathogenic     germline   \n",
       "3                                   Uncertain significance     germline   \n",
       "4        Conflicting interpretations of pathogenicity; ...     germline   \n",
       "...                                                    ...          ...   \n",
       "1207701                                  Likely pathogenic     germline   \n",
       "1207702                             Uncertain significance     germline   \n",
       "1207703                                  Likely pathogenic     germline   \n",
       "1207704                                  Likely pathogenic     germline   \n",
       "1207705                                  Likely pathogenic     germline   \n",
       "\n",
       "        GeneSymbol     HGNC_ID  \\\n",
       "0           ZNF592  HGNC:28986   \n",
       "1          FOXRED1  HGNC:26927   \n",
       "2          FOXRED1  HGNC:26927   \n",
       "3            NUBPL  HGNC:20278   \n",
       "4              HFE   HGNC:4886   \n",
       "...            ...         ...   \n",
       "1207701        GAA   HGNC:4065   \n",
       "1207702     DHTKD1  HGNC:23537   \n",
       "1207703     COL2A1   HGNC:2200   \n",
       "1207704      NAGLU   HGNC:7632   \n",
       "1207705     NOTCH3   HGNC:7883   \n",
       "\n",
       "                                             PhenotypeList  RS# (dbSNP)  \n",
       "0                                Galloway-Mowat syndrome 1    150829393  \n",
       "1        Mitochondrial complex 1 deficiency, nuclear ty...    267606829  \n",
       "2        Mitochondrial complex 1 deficiency, nuclear ty...    267606830  \n",
       "3        not provided|Inborn genetic diseases|Mitochond...    200401432  \n",
       "4        Hemochromatosis type 1|Porphyria cutanea tarda...      1800562  \n",
       "...                                                    ...          ...  \n",
       "1207701                  Glycogen storage disease, type II           -1  \n",
       "1207702         Charcot-Marie-Tooth disease axonal type 2Q    778723208  \n",
       "1207703             Spondyloepiphyseal dysplasia congenita           -1  \n",
       "1207704                   Mucopolysaccharidosis, MPS-III-B           -1  \n",
       "1207705  Cerebral arteriopathy, autosomal dominant, wit...           -1  \n",
       "\n",
       "[2507797 rows x 8 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_processed_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a76f96d2-8bb8-480e-893b-b36cb32cecf3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
